Add blog article to announce Secret references in NodeExpandVolume (#33979)
kubernetes/website: content/en/blog/_posts/2022-04-19-csi-nodeexpandsecret.md

---
layout: blog
title: "Kubernetes 1.25: Use secrets while expanding csi volumes on Node"
date: 2022-07-26T10:00:00-08:00

> slug: kubernetes-1-24-use-secrets-while-expanding-csivolumes-on-node-alpha

Reviewer: I get that we don't have shrink yet. However, this is the mechanism we'd use to pass credentials for a shrink, right?

Imagine that we do add shrink and then people want to find out how to configure secrets, only the article about this makes out that the secrets are only for expand.

In the _text_ of the article we should absolutely make it clear by the choice of words that expand is the action being discussed.

---

**Author:** Humble Chirammal (Red Hat), Louis Koo (deeproute.ai)

Kubernetes v1.25, released earlier this month, introduced a new feature
that lets your cluster expand storage volumes, even when access to those
volumes requires a secret (for example: a credential for accessing a SAN fabric)
to perform node expand operation. This new behavior is in alpha and you
must enable a feature gate (`CSINodeExpandSecret`) to make use of it.
you must also be using [CSI](https://kubernetes-csi.github.io/docs/)
storage; this change isn't relevant to storage drivers that are built in to Kubernetes.

> under the feature gate  (CSINodeExpandSecret) which once enabled in the

Reviewer 2: ```suggestion
under the feature gate `CSINodeExpandSecret` which once enabled in the
```

kube api server and kubelet, provide a mechanism to receive secretRef

> configuration as part of NodeExpansion  by the CSI drivers thus make use of

Author: it seems to me that, above explanation is bit deviated from the message we are trying to convey.

operations that take place when the storage was mounted onto a node and when
when the volume has to be expanded at node side.
Even-though Kubernetes CSI have implemented similar mechanism for Controller
side operations, ie secretRef field available in the csi PV source and making
use of it while controllerExpand request has been sent to the CSI driver, similar
field was missing in the nodeExpandVolume request.  This was causing some trouble
for different CSI drivers to fulfill below scenarios while it perform NodeExpansion
operation.

- At times, the CSI driver need to check the actual size of the backend volume/image
  before proceeding on node side FS resize operation to avoid false positive returns
  from the  backend storage cluster on fs resize operation.

- Encrypted device with LUKs, which need the passphrase in order to resize
  the device on the node.

- For various validations at time of node expansion the CSI driver has to be connected
  to the backend storage cluster, if the secretRef is part of the
  nodeExpandVolume request the CSI driver can make use of the same and connect to
  the storage cluster  to perform the cluster operations.

## How does it work?

> To enable this functionality from this version of Kubernetes, we have introduced

Author: I have replaced the `we` to `Kubernetes Storage SIG` .

a new feature gate called CSINodeExpandSecret. Once the feature gate is enabled
in the cluster, the secretRef field will be added to the NodeExpandVolume request
which has been sent from the kubernetes orchestrator to the CSI driver.
The admin can specify these secrets as an opaque parameter in storage class as

> he/she specifies other CSI secrets today.  The secrets mention in the storage class

Reviewer 2: Maybe we can use `they` instead of `he/she`, not sure which is better.

has to carry the key name as shown below

>  `csi.storage.k8s.io/node-expand-secret-name: test-secret
>   csi.storage.k8s.io/node-expand-secret-namespace: default`

Reviewer 2: ````suggestion
```
csi.storage.k8s.io/node-expand-secret-name: test-secret
csi.storage.k8s.io/node-expand-secret-namespace: default`
```
````

If feature gates are enabled and storage class carries the above secret configuration,
the subject CSI provisioner will receive the secrets as part of the NodeExpansion request.

CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
field set. If not set, the NodeExpandVolume CSI RPC call will be made without a secret.

## Trying it out

> Step 1: enable CSINodeExpandSecret feature gates ( Refer featuregate
> documentation (https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/))

Reviewer 2: ```suggestion
Step 1: enable `CSINodeExpandSecret` feature gates (Please refer to
[Feature Gates](/docs/reference/command-line-tools-reference/feature-gates/))
```

Step 2: Secret and storage class creation

Ex: secret.yaml:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: test-secret
data:
  username: bXktYXBw
  password: Mzk1MjgkdmRnN0pi
```

Ex: sc.yaml:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-hostpath-sc
parameters:

>   csi.storage.k8s.io/node-expand-secret-name: test-secret
>   csi.storage.k8s.io/node-expand-secret-namespace: default

Reviewer: Comments help make the important lines stand out. For example:
```suggestion
  csi.storage.k8s.io/node-expand-secret-name: test-secret   #  the name of the Secret
  csi.storage.k8s.io/node-expand-secret-namespace: default  # the namespace that the Secret is in
```

provisioner: hostpath.csi.k8s.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
allowVolumeExpansion: true
```

## Example output:

> If the pvc was created successfully, we can see it in the pv.spec.csi with cmd `kubectl get pv <pv_name> -o yaml`

Reviewer 2: ```suggestion
If the `pvc` was created successfully, we can see it in the `pv.spec.csi` with command `kubectl get pv <pv_name> -o yaml`
```

Reviewer: This needs a bigger rewrite:
- avoid using “we”
- don't start a new 2nd-level section half way through a demo
- make it clear what the reader has to do do get to a point where they can create a working PersistentVolumeClaim
- use long names, eg persistentvolumeclaim not pvc, in example commands.

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: hostpath.csi.k8s.io

>   creationTimestamp: "2022-03-23T15:14:07Z"

Reviewer: (nit) Bump this to be closer to the release date

  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-95eb531a-d675-49f6-940b-9bc3fde83eb0
  resourceVersion: "420263"
  uid: 6fa824d7-8a06-4e0c-b722-d3f897dcbd65
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 6Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: csi-pvc
    namespace: default
    resourceVersion: "419862"
    uid: 95eb531a-d675-49f6-940b-9bc3fde83eb0
  csi:
    driver: hostpath.csi.k8s.io
    nodeExpandSecretRef:
      name: test-secret
      namespace: default
    volumeAttributes:
      storage.kubernetes.io/csiProvisionerIdentity: 1648042783218-8081-hostpath.csi.k8s.io
    volumeHandle: e21c7809-aabb-11ec-917a-2e2e254eb4cf
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: topology.hostpath.csi/node
          operator: In
          values:
          - rook-node01
  persistentVolumeReclaimPolicy: Delete
  storageClassName: csi-hostpath-sc
  volumeMode: Filesystem
status:
  phase: Bound
```

> Expand the volume online, the secret will be pass to the csi driver, like this,
> the logs in host path csi driver:

Reviewer: ```suggestion
If you then trigger online storage expansion, the kubelet passes the appropriate credentials to the CSI driver,
by loading that Secret and passing the data to the storage driver.

Here's an example debug log:
```

> ```yaml

Reviewer: Don't mark the log excerpt as YAML.

> ```yaml
> I0330 03:29:51.966241       1 server.go:101] GRPC call: /csi.v1.Node/NodeExpandVolume
> I0330 03:29:51.966261       1 server.go:105] GRPC request: {"capacity_range":{"required_bytes":7516192768},"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/hostpath.csi.k8s.io/f7c62e6e08ce21e9b2a95c841df315ed4c25a15e91d8fcaf20e1c2305e5300ab/globalmount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":7}},"volume_id":"e21c7809-aabb-11ec-917a-2e2e254eb4cf","volume_path":"/var/lib/kubelet/pods/bcb1b2c4-5793-425c-acf1-47163a81b4d7/volumes/kubernetes.io~csi/pvc-95eb531a-d675-49f6-940b-9bc3fde83eb0/mount"}
> I0330 03:29:51.966360       1 nodeserver.go:459] req:volume_id:"e21c7809-aabb-11ec-917a-2e2e254eb4cf" volume_path:"/var/lib/kubelet/pods/bcb1b2c4-5793-425c-acf1-47163a81b4d7/volumes/kubernetes.io~csi/pvc-95eb531a-d675-49f6-940b-9bc3fde83eb0/mount" capacity_range:<required_bytes:7516192768 > staging_target_path:"/var/lib/kubelet/plugins/kubernetes.io/csi/hostpath.csi.k8s.io/f7c62e6e08ce21e9b2a95c841df315ed4c25a15e91d8fcaf20e1c2305e5300ab/globalmount" volume_capability:<mount:<> access_mode:<mode:SINGLE_NODE_MULTI_WRITER > > secrets:<key:"XXXXXX" value:"XXXXX" > secrets:<key:"XXXXX" value:"XXXXXX" >
> ```

Reviewer 2: ````suggestion
```console
I0330 03:29:51.966241       1 server.go:101] GRPC call: /csi.v1.Node/NodeExpandVolume
I0330 03:29:51.966261       1 server.go:105] GRPC request: {"capacity_range":{"required_bytes":7516192768},"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/hostpath.csi.k8s.io/f7c62e6e08ce21e9b2a95c841df315ed4c25a15e91d8fcaf20e1c2305e5300ab/globalmount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":7}},"volume_id":"e21c7809-aabb-11ec-917a-2e2e254eb4cf","volume_path":"/var/lib/kubelet/pods/bcb1b2c4-5793-425c-acf1-47163a81b4d7/volumes/kubernetes.io~csi/pvc-95eb531a-d675-49f6-940b-9bc3fde83eb0/mount"}
I0330 03:29:51.966360       1 nodeserver.go:459] req:volume_id:"e21c7809-aabb-11ec-917a-2e2e254eb4cf" volume_path:"/var/lib/kubelet/pods/bcb1b2c4-5793-425c-acf1-47163a81b4d7/volumes/kubernetes.io~csi/pvc-95eb531a-d675-49f6-940b-9bc3fde83eb0/mount" capacity_range:<required_bytes:7516192768 > staging_target_path:"/var/lib/kubelet/plugins/kubernetes.io/csi/hostpath.csi.k8s.io/f7c62e6e08ce21e9b2a95c841df315ed4c25a15e91d8fcaf20e1c2305e5300ab/globalmount" volume_capability:<mount:<> access_mode:<mode:SINGLE_NODE_MULTI_WRITER > > secrets:<key:"XXXXXX" value:"XXXXX" > secrets:<key:"XXXXX" value:"XXXXXX" >
```
````

## The future

As this feature is still in alpha, we expect to update or get feedback from CSI driver
authors with more tests and implementation. The community plans to eventually
promote the feature to Beta in upcoming releases.

> ## How can I learn more?

Reviewer: ```suggestion
## Get involved or learn more
```
“I” doesnt make sense. You'd be switching from “we” representing the authors / contributors to “I” representing the reader, which jars.

The enhancement proposal includes lots of detail about the history and technical
implementation of this feature.

> To know more about Storage Class based dynamic provisioning in Kubernetes please refer:
> [Storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/)
> [Persistent Volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)

Reviewer 2: ```suggestion
To know more about Storage Class based dynamic provisioning in Kubernetes please refer to:
[Storage class](/docs/concepts/storage/storage-classes/),
[Persistent Volumes](/docs/concepts/storage/persistent-volumes/)
```

Reviewer: ```suggestion
To know more about StorageClass-based dynamic provisioning in Kubernetes, please refer to:
[Storage Classes](/docs/concepts/storage/storage-classes/) and
[Persistent Volumes](/docs/concepts/storage/persistent-volumes/).
```

> Please get involved by joining the Kubernetes storage SIG (https://github.com/kubernetes/community/blob/master/sig-storage/README.md) to help us enhance this feature. There are a lot of good ideas already and we'd be thrilled to have more!

Reviewer 2: ```suggestion
Please get involved by joining the [Kubernetes storage SIG](https://github.com/kubernetes/community/blob/master/sig-storage/README.md) to help us enhance this feature. There are a lot of good ideas already and we'd be thrilled to have more!
```

Reviewer: ```suggestion
Please get involved by joining the Kubernetes [Storage SIG](https://github.com/kubernetes/community/blob/master/sig-storage/README.md) (Special Interest Group) to help us enhance this feature. There are a lot of good ideas already and we'd be thrilled to have more!
```

---
layout: blog
title: "Kubernetes 1.25: Use Secrets for Node-Driven Resize of CSI Volumes."
date: 2022-07-26T10:00:00-08:00
slug: kubernetes-1-25-use-secrets-while-expanding-csi-volumes-on-node-alpha
---

**Author:** Humble Chirammal (Red Hat), Louis Koo (deeproute.ai)

Kubernetes v1.25, released earlier this month, introduced a new feature
that lets your cluster expand storage volumes, even when access to those
volumes requires a secret (for example: a credential for accessing a SAN fabric)
to perform node expand operation. This new behavior is in alpha and you
must enable a feature gate (`CSINodeExpandSecret`) to make use of it.
You must also be using [CSI](https://kubernetes-csi.github.io/docs/)
storage; this change isn't relevant to storage drivers that are built in to Kubernetes.

TO turn on this new, alpha feature, you enable the `CSINodeExpandSecret` feature
gate for the kube api server and kubelet, provide a mechanism to receive secretRef
configuration as part of NodeExpansion  by the CSI drivers thus make use of
the same to perform node side expansion operation with the underlying

> storage system.

Reviewer: I'm worried that readers who have only read up to this points won't know what the new feature does. I'll try to suggest some different wording here.

a new feature gate called `CSINodeExpandSecret`. Once the feature gate is enabled
in the cluster, NodeExpandVolume requests can include a `secretRef` field. The NodeExpandVolume request
is part of CSI; for example, in a request which has been sent from the Kubernetes
control plane to the CSI driver.
The admin can specify these secrets as an opaque parameter in storage class as
he/she specifies other CSI secrets today.  The secrets mention in the storage class
has to carry the key name as shown below

```
csi.storage.k8s.io/node-expand-secret-name: test-secret
csi.storage.k8s.io/node-expand-secret-namespace: default`
```

If feature gates are enabled and storage class carries the above secret configuration,
the subject CSI provisioner receives the credentials from the Secret as part of the NodeExpansion request.

CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
field set. If not set, the NodeExpandVolume CSI RPC call will be made without a secret.

## Trying it out

Step 1: enable `CSINodeExpandSecret` feature gate (Please refer to
[Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/))

> Step 2: Secret and storage class creation

Reviewer: I'd write
1. Enable the `CSINodeExpandSecret` feature gate (please refer to
[Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)).

1. Create a Secret, and then a StorageClass that uses that Secret
```
(Markdown does automatic numbering)

> Ex: Here's an example manifest for a Secret that holds credentials:

Reviewer: ```suggestion
Here's an example manifest for a Secret that holds credentials:
```

---
layout: blog

> title: "Kubernetes 1.25: Use Secrets for Node-Driven Expansion of CSI Volumes."

Reviewer: I'd do this:
```suggestion
title: >-
  Kubernetes 1.25: Use Secrets for Node-Driven Expansion of CSI Volumes
```

The main change is to omit the final period.

> date: 2022-07-26T10:00:00-08:00

Reviewer: The release comms team can assign an actual publication date, possibly just a provisional one for now.

Author: Yeah, put it as 26th Aug for now.

The Kubernetes CSI already implemented a similar mechanism specific kinds of
volume resizes; namely, resizes of PersistentVolumes where the resizes take place
independently from any node. In that case, you associate a PersistentVolume
with a Secret that contains credentials for volume resize actions.
CSI also supports a `nodeExpandVolume` operation, where the resize is driven
from a node in your cluster where the volume is attached. Having resize (expansion)
operations sometimes able to use credentials and sometimes not meant it was difficult
for people to design CSI drivers that could cover all resize scenarios.

- At times, the CSI driver needs to check the actual size of the backend block storage (or image)
  before proceeding with a node-level filesystem expand operation. This avoids false positive returns
  from the backend storage cluster during filesystem expands.

- When a PersistentVolume represents encrypted block storage (for example using LUKS)
  you need to provide a passphrase in order to expand the device, and also to make it possible
  to grow the filesystem on that device.

- For various validations at time of node expansion, the CSI driver has to be connected

>   to the backend storage cluster. If the` nodeExpandVolume` request includes a `secretRef`

Reviewer: ```suggestion
  to the backend storage cluster. If the `nodeExpandVolume` request includes a `secretRef`
```

  then the CSI driver can make use of the same and connect to the storage cluster to
  perform the cluster operations.

## How does it work?

> To enable this functionality from this version of Kubernetes, Kubernetes Storage SIG have introduced

Reviewer: ```suggestion
To enable this functionality from this version of Kubernetes, SIG Storage have introduced
```

a new feature gate called `CSINodeExpandSecret`. Once the feature gate is enabled
in the cluster, NodeExpandVolume requests can include a `secretRef` field. The NodeExpandVolume request
is part of CSI; for example, in a request which has been sent from the Kubernetes
control plane to the CSI driver.

> The admin can specify these secrets as an opaque parameter in storage class as
> they specifies other CSI secrets today.  The secrets mention in the storage class
> has to carry the key name as shown below

Reviewer: Is this right?
```suggestion
As a cluster operator, you admin can specify these secrets as an opaque parameter in a StorageClass,
the same way that you can already specify other CSI secret data. The StorageClass needs to have some
CSI-specific parameters set. Here's an example of those parameters:
```

(these look like annotations or labels, not fields)

Author: Yeah.. these are parameters in SC.. but the format is like annotations though.

```
csi.storage.k8s.io/node-expand-secret-name: test-secret

> csi.storage.k8s.io/node-expand-secret-namespace: default`

Reviewer: ```suggestion
csi.storage.k8s.io/node-expand-secret-namespace: default
```

```

If feature gates are enabled and storage class carries the above secret configuration,
the subject CSI provisioner receives the credentials from the Secret as part of the NodeExpansion request.

CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
field set. If not set, the NodeExpandVolume CSI RPC call will be made without a secret.

## Trying it out

1. Enable the `CSINodeExpandSecret` feature gate (please refer to
[Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)).

> 1. Create a Secret, and then a StorageClass that uses that Secret

Reviewer: (nit)
```suggestion
1. Create a Secret, and then a StorageClass that uses that Secret.
```

Here's an example manifest for a Secret that holds credentials:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: test-secret
  namespace: default
data:
stringData:
  username: admin
  password: t0p-Secret
```

Here's an example manifest for a StorageClass that refers to those credentials:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-hostpath-sc
parameters:
  csi.storage.k8s.io/node-expand-secret-name: test-secret   # the name of the Secret
  csi.storage.k8s.io/node-expand-secret-namespace: default  # the namespace that the Secret is in

> provisioner: hostpath.csi.k8s.io

Reviewer: Does this provisioner actually work with NodeExpandVolume secret references?

(if not, maybe make up a fictional provisioner, eg `blockstorage.cloudprovider.example`?)

Author: yeah.. the hostpath output and logs are from real testing. should be fine. :+1:

Reviewer: How does the `hostPath` provisioner actually _use_ these secrets? If it ignores them, that's maybe not an ideal example.

Author: actually this feature is from the kubernetes csi client to push the secrets to the CSI driver as part of the NodeExpandVolume call.. The hostPath provisioner is a CSI driver who get those as shown in the example.. in that sense, it looks fine to me keep the current version as an example .. @sftim

Author: if really needed I can replace the string `hostpath` to `hostpath-example` or something similar.. please let me know wdyt @sftim

Reviewer: Could you explain in a bit more detail (to reviewers, not readers) how a cluster admin works out what they put into `test-secret`? If I changed the contents of that Secret to have empty strings as the values, would it still work?

If so, then we're not really illustrating things end-to-end. Yes, the _feature_ from the KEP is implemented, but the reader cares about the end-to-end story. They're less interested in seeing an integration test that proves that the secret reference gets to the CSI driver.

Is `blockstorage.cloudprovider.example` a plausible CSI driver name / reference? If it is, how about using that?

Author: @sftim the nodeExpandSecret follow the same convention or rules of other similar secrets ( for ex: node stage, publish, controller expand) in the SC as described here 
https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html.  The secret values will be sent to the CSI driver and if the credentials are correct to connect or perform the requested operation it will work otherwise fail. 

>Is blockstorage.cloudprovider.example a plausible CSI driver name / reference? If it is, how about using that?
Sure, I have changed the provisioner name to above.. it sound fine to have above as a pluasible CSI driver name.. 
ptal.. thanks.

reclaimPolicy: Delete
volumeBindingMode: Immediate
allowVolumeExpansion: true
```

## Example output

If the PersistentVolumeClaim (PVC) was created successfully, you can see that
configuration within the `spec.csi` field of the PersistentVolume (look for
`spec.csi.nodeExpandSecretRef`).
Check that it worked by running `kubectl get persistentvolume <pv_name> -o yaml`.
You should see something like.

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: hostpath.csi.k8s.io

>   creationTimestamp: "2022-07-23T15:14:07Z"

Reviewer: (nit)
Change this date to be after the release but before the article publishes.

Author: sure put the date after 23rd Aug and before the blog publishing provisioner date of 28 Aug. :+1:

Author: Thanks @sftim

I0330 03:29:51.966360       1 nodeserver.go:459] req:volume_id:"e21c7809-aabb-11ec-917a-2e2e254eb4cf" volume_path:"/var/lib/kubelet/pods/bcb1b2c4-5793-425c-acf1-47163a81b4d7/volumes/kubernetes.io~csi/pvc-95eb531a-d675-49f6-940b-9bc3fde83eb0/mount" capacity_range:<required_bytes:7516192768 > staging_target_path:"/var/lib/kubelet/plugins/kubernetes.io/csi/hostpath.csi.k8s.io/f7c62e6e08ce21e9b2a95c841df315ed4c25a15e91d8fcaf20e1c2305e5300ab/globalmount" volume_capability:<mount:<> access_mode:<mode:SINGLE_NODE_MULTI_WRITER > > secrets:<key:"XXXXXX" value:"XXXXX" > secrets:<key:"XXXXX" value:"XXXXXX" >
```

## The future

As this feature is still in alpha, Kubernetes Storage SIG expect to update or get feedback from CSI driver
authors with more tests and implementation. The community plans to eventually
promote the feature to Beta in upcoming releases.

## Get involved or learn more?

The enhancement proposal includes lots of detail about the history and technical
implementation of this feature.

> To know more about Storage Class based dynamic provisioning in Kubernetes please refer:

Reviewer: (nit)
```suggestion
To learn more about StorageClass based dynamic provisioning in Kubernetes, please refer to
```

> ---

Reviewer 2: The file name `2022-04-19-csi-nodeexpandsecret.md` should be changed.

layout: blog
title: >-
  Kubernetes 1.25: Use Secrets for Node-Driven Expansion of CSI Volumes

> date: 2022-08-28T10:00:00-08:00

Reviewer 2: In 420 blog posts, only 6 of them use `2022-08-02T12:00:00+08:00` format.
412 of them using `date: 2022-08-02` format.

```suggestion
date: 2022-08-28
```

## How does it work?

To enable this functionality from this version of Kubernetes, SIG Storage have introduced
a new feature gate called `CSINodeExpandSecret`. Once the feature gate is enabled
in the cluster, NodeExpandVolume requests can include a `secretRef` field. The NodeExpandVolume request
is part of CSI; for example, in a request which has been sent from the Kubernetes
control plane to the CSI driver.

As a cluster operator, you admin can specify these secrets as an opaque parameter in a StorageClass,
the same way that you can already specify other CSI secret data. The StorageClass needs to have some
CSI-specific parameters set. Here's an example of those parameters:

```
csi.storage.k8s.io/node-expand-secret-name: test-secret
csi.storage.k8s.io/node-expand-secret-namespace: default
```

If feature gates are enabled and storage class carries the above secret configuration,
the subject CSI provisioner receives the credentials from the Secret as part of the NodeExpansion request.

CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
field set. If not set, the NodeExpandVolume CSI RPC call will be made without a secret.

## Trying it out

> 1. Enable the `CSINodeExpandSecret` feature gate (please refer to
> [Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)).

Reviewer 2: ```suggestion
1. Enable the `CSINodeExpandSecret` feature gate (please refer to
   [Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)).
```

```

## The future

As this feature is still in alpha, Kubernetes Storage SIG expect to update or get feedback from CSI driver
authors with more tests and implementation. The community plans to eventually
promote the feature to Beta in upcoming releases.

## Get involved or learn more?

The enhancement proposal includes lots of detail about the history and technical
implementation of this feature.

To learn more about StorageClass based dynamic provisioning in Kubernetes, please refer to
[Storage Classes](/docs/concepts/storage/storage-classes/) and
[Persistent Volumes](/docs/concepts/storage/persistent-volumes/).

> Please get involved by joining the Kubernetes [Storage SIG](https://github.com/kubernetes/community/blob/master/sig-storage/README.md) (Special Interest Group) to help us enhance this feature. There are a lot of good ideas already and we'd be thrilled to have more!

Reviewer 2: We can break into few lines like this:

```suggestion
Please get involved by joining the Kubernetes
[Storage SIG](https://github.com/kubernetes/community/blob/master/sig-storage/README.md)
(Special Interest Group) to help us enhance this feature.
There are a lot of good ideas already and we'd be thrilled to have more!
```
